src/snowflake/snowpark/_internal/packaging_utils.py

```diff
@@ -0,0 +1,251 @@
+    Python package, finds the name of the package.
+    This is found on a line containing "Name: my_package".
+
+    Args:
+        metadata_file_path (str): The path to the METADATA file
+
+    Returns:
+        str: the name of the package.
+    """
+    import re
+
+    with open(metadata_file_path, encoding="utf-8") as metadata_file:
+        contents = metadata_file.read()
+        results = re.search("^Name: (.*)$", contents, flags=re.MULTILINE)
+        if results is None:
+            return None
+        requirement_line = results.group(1)
+        results = re.search("^Version: (.*)$", contents, flags=re.MULTILINE)
+        if results is not None:
+            version = results.group(1)
+            requirement_line += f"=={version}"
+        return requirement_line.strip().lower()
+
+
+def get_downloaded_packages(directory: str) -> Dict[Requirement, List[str]]:
+    import glob
+    import os
+
+    metadata_files = glob.glob(os.path.join(directory, "*dist-info", "METADATA"))
+    return_dict: Dict[Requirement, List[str]] = {}
+    for metadata_file in metadata_files:
+        parent_folder = os.path.dirname(metadata_file)
+        package = get_package_name_from_metadata(metadata_file)
+
+        if package is not None:
+            # Determine which folders or files belong to this package
+            record_file_path = os.path.join(parent_folder, "RECORD")
+            if os.path.exists(record_file_path):
+                # Get unique root folder names
+                with open(record_file_path, encoding="utf-8") as record_file:
+                    record_entries = list(
+                        {
+                            os.path.split(line)[0].split(",")[0]
+                            for line in record_file.readlines()
+                        },
+                    )
+                    included_record_entries = []
+                    for record_entry in record_entries:
+                        record_entry_full_path = os.path.abspath(
+                            os.path.join(directory, record_entry),
+                        )
+                        # RECORD file might contain relative paths to items outside target folder. (ignore these)
+                        if (
+                            os.path.exists(record_entry_full_path)
+                            and directory in record_entry_full_path
+                            and len(record_entry) > 0
+                        ):
+                            included_record_entries.append(record_entry)
+                    package_req = Requirement.parse(package)
+                    return_dict[package_req] = included_record_entries
+    return return_dict
+
+
+def identify_supported_packages(
+    packages: List[Requirement],
+    valid_packages: Dict[str, List[str]],
+    native_packages: Set[str],
+) -> Tuple[List[Requirement], List[Requirement], List[Requirement]]:
+    supported_dependencies: List = []
```

Comment:
Let's document what the function does and the meaning of the parameters. It'll be great if we can briefly describe what valid/native packages are.